# Credit Card Customer Churn Analysis – ETL Process

The dataset contains customer information from a bank, including credit card usage, demographic details, account activity, and a churn indicator.

The objective of this ETL process is to:
- Load and inspect the dataset
- Clean and transform the data
- Prepare the dataset for analysis and visualisation


In [2]:
import pandas as pd

**1. Import Required Libraries**

We begin by importing the necessary Python libraries for data handling and exploration.


In [3]:
import pandas as pd
import numpy as np


**2. Load the Dataset**

We load the CSV file named `BankChurners.csv` from the local directory. 


In [4]:
#Step 1. Load the dataset
file_path = r'C:\Users\Sheila\Documents\VSCODE PROJECTS\Hackathon2_Credit_Card_Churn_T5\Data\Raw\BankChurners.csv'
 
df = pd.read_csv(file_path)

# Step 2. Show the first 5 rows
df.head()


CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

**3. Clean the Dataset**

We remove irrelevant columns, rename important ones for clarity, and convert categorical values for analysis.

In this step, we remove:
- `CLIENTNUM`: a unique identifier not useful for analysis
- `Unnamed: 0`: an index column accidentally included
- Change the label of Naive_Bayes_Classifier_Attrition_Flag_No and Naive_Bayes_Classifier_Attrition_Flag_Yes to a shorter label which will be Naive_Bayes_1 and Naive_Bayes_2 respectively.


We also:
- Rename the churn indicator column for clarity
- Convert churn status into binary values: 1 = Attrited, 0 = Existing




In [5]:
#Step 3. Drop irrelevant or auto-generated columns and rename Naive Bayes columns.
columns_to_drop = ['CLIENTNUM', 'Unnamed: 0']
#Step 4. Identify the two columns starting with 'Naive_Bayes_Classifier'
naive_bayes_cols = [col for col in df.columns if col.startswith('Naive_Bayes_Classifier')]

#Step 5. Rename them safely to 'Naive_Bayes_1' and 'Naive_Bayes_2'
if len(naive_bayes_cols) == 2:
    df.rename(columns={
        naive_bayes_cols[0]: 'Naive_Bayes_1',
        naive_bayes_cols[1]: 'Naive_Bayes_2'
    }, inplace=True)
else:
    print(f"Expected 2 columns, found {len(naive_bayes_cols)}: {naive_bayes_cols}")


#Step 6. Drop columns safely (ignore if not found)
df.drop(columns=columns_to_drop, errors='ignore', inplace=True)

#Step 7. Rename churn column
df.rename(columns={'Attrition_Flag': 'Churn_Status'}, inplace=True)

#Step 8. Encode churn status
df['Churn_Status'] = df['Churn_Status'].map({'Attrited Customer': 1, 'Existing Customer': 0})

# Step 9. Show the first 5 rows
df.head()

Churn_Status  Customer_Age Gender  Dependent_count Education_Level  \
0             0            45      M                3     High School   
1             0            49      F                5        Graduate   
2             0            51      M                3        Graduate   
3             0            40      F                4     High School   
4             0            40      M                3      Uneducated   

  Marital_Status Income_Category Card_Category  Months_on_book  \
0        Married     $60K - $80K          Blue              39   
1         Single  Less than $40K          Blue              44   
2        Married    $80K - $120K          Blue              36   
3        Unknown  Less than $40K          Blue              34   
4        Married     $60K - $80K          Blue              21   

   Total_Relationship_Count  ...  Credit_Limit  Total_Revolving_Bal  \
0                         5  ...       12691.0                  777   
1                         6  ...        8256.0                  864   
2                         4  ...        3418.0                    0   
3                         3  ...        3313.0                 2517   
4                         5  ...        4716.0                    0   

   Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
0          11914.0                 1.335             1144              42   
1           7392.0                 1.541             1291              33   
2           3418.0                 2.594             1887              20   
3            796.0                 1.405             1171              20   
4           4716.0                 2.175              816              28   

   Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  Naive_Bayes_1  Naive_Bayes_2  
0                1.625                  0.061       0.000093        0.99991  
1                3.714                  0.105       0.000057        0.99994  
2                2.333                  0.000       0.000021        0.99998  
3                2.333                  0.760       0.000134        0.99987  
4                2.500                  0.000       0.000022        0.99998  

[5 rows x 22 columns]

**4. Generate Complete Descriptive Statistics Table**

This expanded statistical summary includes:
- Central tendency (mean, median)
- Dispersion (standard deviation, min, max, IQR)
- Distribution shape (skewness, kurtosis)
- Missing value count

These insights help detect outliers, asymmetry, and data quality issues before modelling or visualisation.


In [6]:
import numpy as np

#Step 8. Select only numerical columns
num_df = df.select_dtypes(include=[np.number])

#Step 9. Build custom descriptive statistics
stats_df = pd.DataFrame()

stats_df['count'] = num_df.count()
stats_df['mean'] = num_df.mean()
stats_df['std'] = num_df.std()
stats_df['min'] = num_df.min()
stats_df['25%'] = num_df.quantile(0.25)
stats_df['median'] = num_df.median()
stats_df['75%'] = num_df.quantile(0.75)
stats_df['max'] = num_df.max()
stats_df['skewness'] = num_df.skew()
stats_df['kurtosis'] = num_df.kurtosis()
stats_df['missing_values'] = num_df.isnull().sum()

#Step 10. Display nicely in Jupyter
display(stats_df.round(3))


count      mean       std     min       25%  \
Churn_Status              10127     0.161     0.367     0.0     0.000   
Customer_Age              10127    46.326     8.017    26.0    41.000   
Dependent_count           10127     2.346     1.299     0.0     1.000   
Months_on_book            10127    35.928     7.986    13.0    31.000   
Total_Relationship_Count  10127     3.813     1.554     1.0     3.000   
Months_Inactive_12_mon    10127     2.341     1.011     0.0     2.000   
Contacts_Count_12_mon     10127     2.455     1.106     0.0     2.000   
Credit_Limit              10127  8631.954  9088.777  1438.3  2555.000   
Total_Revolving_Bal       10127  1162.814   814.987     0.0   359.000   
Avg_Open_To_Buy           10127  7469.140  9090.685     3.0  1324.500   
Total_Amt_Chng_Q4_Q1      10127     0.760     0.219     0.0     0.631   
Total_Trans_Amt           10127  4404.086  3397.129   510.0  2155.500   
Total_Trans_Ct            10127    64.859    23.473    10.0    45.000   
Total_Ct_Chng_Q4_Q1       10127     0.712     0.238     0.0     0.582   
Avg_Utilization_Ratio     10127     0.275     0.276     0.0     0.023   
Naive_Bayes_1             10127     0.160     0.365     0.0     0.000   
Naive_Bayes_2             10127     0.840     0.365     0.0     1.000   

                            median        75%        max  skewness  kurtosis  \
Churn_Status                 0.000      0.000      1.000     1.848     1.417   
Customer_Age                46.000     52.000     73.000    -0.034    -0.289   
Dependent_count              2.000      3.000      5.000    -0.021    -0.683   
Months_on_book              36.000     40.000     56.000    -0.107     0.400   
Total_Relationship_Count     4.000      5.000      6.000    -0.162    -1.006   
Months_Inactive_12_mon       2.000      3.000      6.000     0.633     1.099   
Contacts_Count_12_mon        2.000      3.000      6.000     0.011     0.001   
Credit_Limit              4549.000  11067.500  34516.000     1.667     1.809   
Total_Revolving_Bal       1276.000   1784.000   2517.000    -0.149    -1.146   
Avg_Open_To_Buy           3474.000   9859.000  34516.000     1.662     1.799   
Total_Amt_Chng_Q4_Q1         0.736      0.859      3.397     1.732     9.994   
Total_Trans_Amt           3899.000   4741.000  18484.000     2.041     3.894   
Total_Trans_Ct              67.000     81.000    139.000     0.154    -0.367   
Total_Ct_Chng_Q4_Q1          0.702      0.818      3.714     2.064    15.689   
Avg_Utilization_Ratio        0.176      0.503      0.999     0.718    -0.795   
Naive_Bayes_1                0.000      0.000      1.000     1.849     1.418   
Naive_Bayes_2                1.000      1.000      1.000    -1.849     1.418   

                          missing_values  
Churn_Status                           0  
Customer_Age                           0  
Dependent_count                        0  
Months_on_book                         0  
Total_Relationship_Count               0  
Months_Inactive_12_mon                 0  
Contacts_Count_12_mon                  0  
Credit_Limit                           0  
Total_Revolving_Bal                    0  
Avg_Open_To_Buy                        0  
Total_Amt_Chng_Q4_Q1                   0  
Total_Trans_Amt                        0  
Total_Trans_Ct                         0  
Total_Ct_Chng_Q4_Q1                    0  
Avg_Utilization_Ratio                  0  
Naive_Bayes_1                          0  
Naive_Bayes_2                          0

**5. Insights from Descriptive Statistics Table**
    
  **Credit_Limit & Avg_Open_To_Buy**
    
 - Right-skewed with high kurtosis: These two features are heavily right-skewed, indicating that while most customers have relatively low credit limits or available credit, a small number have unusually high values. This uneven distribution may affect models or visuals, so capping or log transformation could help reduce their impact.

  **Total_Trans_Amt & Total_Ct_Chng_Q4_Q1**
    
 - High positive skewness and kurtosis: Both variables show high positive skewness and kurtosis, meaning there are a few customers with extremely high transaction amounts or rapid changes in transaction count from Q4 to Q1. These outliers could distort averages and should be capped or transformed before modelling or segmentation.

  **Customer_Age**
    
 - Low skewness and kurtosis: The age of customers follows a fairly symmetrical distribution with low skewness and kurtosis, suggesting it is close to normally distributed. This makes it a stable variable to use directly in visualisations or predictive models without requiring transformation.

  **Churn_Status**
 - Binary with ~16.1% churned: This is a binary variable showing whether customers have churned, with approximately 16.1% of the dataset marked as churned. 

  **Dependent_count**

 - Mean = 2.35, Skewness ≈ -0.02, Kurtosis = -0.68
 - Very balanced, almost symmetrical: With a mean of 2.35 and nearly zero skewness, this variable is evenly distributed across customers. Most customers have 1 to 3 dependents, and the data is already well balanced.
   

  **Months_on_book**

 - Mean ≈ 36, Range = 13–56 months, Low skewness
 - Good distribution: Customers have been with the company for between 13 and 56 months, with an average around 36 months. The distribution is balanced with low skewness, indicating a good range of customer tenures without concentration at the extremes.
     

  **Total_Relationship_Count**

 - Mean ≈ 3.8, Skewness = -0.16, Kurtosis = -1.00
 - Slightly left-skewed, but well-distributed: This feature tracks how many products or services a customer uses, with a slight left skew. Despite this, the distribution is generally even.
    

   **Months_Inactive_12_mon**

 - Mean ≈ 2.34, Skewness = 0.63, Kurtosis = 1.1
 - Some customers are inactive for long periods: Customers vary in inactivity duration, with a mean around 2.34 months and moderate positive skew. A few customers are inactive for long stretches, so grouping values into bins (e.g. 0, 1–2, 3+) can simplify visualisations and help highlight behaviour patterns.
    

   **Contacts_Count_12_mon**

 - Mean ≈ 2.45, Skewness ≈ 0.01, Kurtosis ≈ 0
 - Very balanced distribution:The number of customer service interactions over the past year is almost perfectly symmetrical and normally distributed. Most customers contact the service centre between 1 and 3 times, so no transformation or grouping is needed here.
    

   **Total_Revolving_Bal**

 - Mean ≈ 1162, Skewness = -0.15, Kurtosis = -1.14
 - Centered with a healthy spread; small negative skew: Balances that revolve month to month are fairly centred around the mean of 1162, with a small left skew. This suggests most customers keep manageable balances, and the distribution is stable enough for direct use without modification.

   **Total_Amt_Chng_Q4_Q1**

 - Mean = 0.76, Skewness = 1.73, Kurtosis = 9.99
 - Positive skew and heavy tails: This variable captures changes in spending from Q4 to Q1 and is highly positively skewed with extreme kurtosis. This suggests a small group of customers had unusually large changes in spending, and these values should be capped to avoid distortion in trend analysis or models.
    

   **Total_Trans_Ct**

 - Mean ≈ 65, Skewness ≈ 0.15, Kurtosis ≈ -0.37
 - Close to normal distribution: The total number of transactions per customer is fairly normally distributed, with a mean around 65. The slight positive skew and moderate kurtosis indicate this feature is stable and does not require transformation before further use.
    

   **Avg_Utilization_Ratio**

 - Mean ≈ 0.28, Skewness = 0.72, Kurtosis = -0.80
 - Right-skewed: This ratio indicates how much of the credit limit is used on average. It shows moderate right skew, with most customers using a small proportion of their credit. Grouping values into usage levels (e.g., Low <30%, Medium 30–70%, High >70%) can enhance visual analysis and customer profiling.



**6. Outlier Handling Using IQR Capping**

Several numerical features in our dataset showed significant skewness and heavy tails in the descriptive statistics. These outliers can distort visualisations, mislead interpretation.
Thus,  we will apply **IQR capping** to limit extreme values but we will preserve the overall distribution of the data.

 The main reason to apply this method is to:
- Reduces the influence of outliers on charts such as boxplots, histograms, and scatter plots.
- Makes distributions more compact and interpretable.
-

Features that will be capped:
- **Credit_Limit**
- **Avg_Open_To_Buy**
- **Total_Trans_Amt**
- **Total_Amt_Chng_Q4_Q1**
- **Total_Ct_Chng_Q4_Q1**
- **Avg_Utilization_Ratio**



This transformation ensures that these columns are now **suitable for visualisations**, producing clearer and more accurate insights.


Before applying  transformation we will count the exact number of outliers.


In [7]:
#Step 11. Define IQR capping function
def cap_outliers_iqr(series):
    Q1 = series.quantile(0.25)
    Q3 = series.quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    return series.clip(lower, upper)

#Step 12. List of features to cap
features_to_cap = [
    'Avg_Open_To_Buy',
    'Total_Trans_Amt',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Ct_Chng_Q4_Q1',
    'Avg_Utilization_Ratio',
    "Credit_Limit"
]

#Step 13. Count outliers before capping
def count_outliers_iqr(series):
    Q1 = series.quantile(0.25)
    Q3 = series.quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    return ((series < lower) | (series > upper)).sum()

total_outliers = 0
for col in features_to_cap:
    count = count_outliers_iqr(df[col])
    print(f"{col}: {count} outliers")
    total_outliers += count

print(f"\nTotal number of outliers before capping: {total_outliers}")

# Step 14. Apply IQR capping to those features
for col in features_to_cap:
    df[col] = cap_outliers_iqr(df[col])



Avg_Open_To_Buy: 963 outliers
Total_Trans_Amt: 896 outliers
Total_Amt_Chng_Q4_Q1: 396 outliers
Total_Ct_Chng_Q4_Q1: 394 outliers
Avg_Utilization_Ratio: 0 outliers
Credit_Limit: 984 outliers

Total number of outliers before capping: 3633


**7. Save the Cleaned Dataset to the Absolute Path**

We save the transformed dataset to the `Input` folder for further analysis or dashboarding.


In [8]:
import os

input_path = r'C:\Users\Sheila\Documents\VSCODE PROJECTS\Hackathon2_Credit_Card_Churn_T5\Data\Input\Cleaned_BankChurners.csv'

#Step 23. Ensure the output directory exists
os.makedirs(os.path.dirname(input_path), exist_ok=True)

#Step 24. Save the cleaned dataset
df.to_csv(input_path, index=False)

print(f"Cleaned dataset saved to: {input_path}")


Cleaned dataset saved to: C:\Users\Sheila\Documents\VSCODE PROJECTS\Hackathon2_Credit_Card_Churn_T5\Data\Input\Cleaned_BankChurners.csv
